### PR TITLE
Add Version Dropdown To 0x.js Docs

### PR DESCRIPTION
**Pull request by fabioberger**
This PR:
- Adds versions to the `/docs/0xjs` URL so users can reference a specific versions docs.
- Adds a version dropdown so users can browse to any given 0x.js version docs.
- Pulls the doc JSON for a specific version from S3, and defaults to the latest version if none specified in URL.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -45,13 +45,15 @@
...
     "scroll-to-element": "^2.0.0",
+    "semver-sort": "0.0.4",
     "thenby": "^1.2.3",
     "truffle-contract": "2.0.1",
     "tslint-config-0xproject": "^0.0.2",
     "uglify-es": "^3.0.18",
     "web3": "^0.18.2",
     "web3-provider-engine": "^11.0.0",
-    "whatwg-fetch": "^2.0.3"
+    "whatwg-fetch": "^2.0.3",
```
Comment (LogvinovLeon) [inline package.json]: Which browsers does this polyfill support that don't have native implementations and that we care about?
fetch has 74% of native support and is not supported only by old safari, IE and Opera Mini.
http://caniuse.com/#feat=fetch
At the same moment this polyfill doesn't claim to be tested on those browsers.
https://github.com/github/fetch#browser-support
So it gives us around additional 6% for the cost of a bigger bundle size.
Can we import it conditionally?

Comment (LogvinovLeon) [inline package.json]: Can we also research the bundle size implications?

File: ts/globals.d.ts [hunk 2]
```diff
@@ -13,9 +13,11 @@ declare module 'react-html5video';
 declare module 'web3-provider-engine/subproviders/filters';
 declare module 'web3-provider-engine/subproviders/rpc';
 declare module 'thenby';
+declare module 'semver-sort';
```
Comment (LogvinovLeon) [inline ts/globals.d.ts]: You must be kidding me ;)

File: ts/globals.d.ts [hunk 3]
```diff
@@ -13,9 +13,11 @@ declare module 'react-html5video';
...
 declare module 'find-versions';
 declare module 'compare-versions';
 declare module 'react-highlight';
+declare module 'xml-js';
```
Comment (LogvinovLeon) [inline ts/globals.d.ts]: same

File: ts/pages/documentation/zero_ex_js_documentation.tsx [hunk 4]
```diff
@@ -321,17 +344,72 @@ export class ZeroExJSDocumentation extends React.Component<ZeroExJSDocumentation
...
+        }
+        const responseXML = await response.text();
+        const responseJSONString: any = convert.xml2json(responseXML, {
+            compact: true,
+        });
+        const responseObj = JSON.parse(responseJSONString);
+        const fileObjs = responseObj.ListBucketResult.Contents;
+        const versionFileNames = _.map(fileObjs, (fileObj: any) => {
+            return fileObj.Key._text;
+        });
```
Comment (LogvinovLeon) [inline ts/pages/documentation/zero_ex_js_documentation.tsx]: Can we refactor getting versionFileNames into separate function?

File: ts/pages/documentation/zero_ex_js_documentation.tsx [hunk 5]
```diff
@@ -321,17 +344,72 @@ export class ZeroExJSDocumentation extends React.Component<ZeroExJSDocumentation
...
+        const versionToFileName: {[version: string]: string} = {};
+        _.each(versionFileNames, fileName => {
+            const version = findVersions(fileName);
```
Comment (LogvinovLeon) [inline ts/pages/documentation/zero_ex_js_documentation.tsx]: doesn't findVersions return an array?
maybe: `const [version] = findVersions(fileName);`?
That's why we need to type this module.

File: ts/pages/documentation/zero_ex_js_documentation.tsx [hunk 6]
```diff
@@ -321,17 +344,72 @@ export class ZeroExJSDocumentation extends React.Component<ZeroExJSDocumentation
...
+        });
+    }
+    private async getJSONDocFileAsync(fileName: string) {
+        const endpoint = `${constants.S3_DOCUMENTATION_JSON_ROOT}/${fileName}`;
+        const response = await fetch(endpoint);
+        if (response.status !== 200) {
+            // TODO: Show the user an error message when the docs fail to load
+            return;
         }
+        const responseText = await response.text();
```
Comment (LogvinovLeon) [inline ts/pages/documentation/zero_ex_js_documentation.tsx]: response.json()?

**Comment by LogvinovLeon** [comment]
Loogk good overall. Can we also cache (in memory) alread loaded jsons, so that switching back to previously viewed version is instant?

File: package.json (hunk 1 above)
Comment (fabioberger) [inline package.json]: Fetch is 13kb, I don't think worrying about perf here makes sense.

File: ts/globals.d.ts (hunk 2 above)
Comment (fabioberger) [inline ts/globals.d.ts]: Haha, thought the same thing but why re-invent the wheel. I'd also say that I wouldn't have imported a huge library that does 100 things to use one method but this is a very well scoped module that does one thing, and one thing well.

File: ts/globals.d.ts (hunk 3 above)
Comment (fabioberger) [inline ts/globals.d.ts]: Same what?

File: ts/globals.d.ts (hunk 2 above)
Comment (LogvinovLeon) [inline ts/globals.d.ts]: I was talking about defining the type

File: ts/globals.d.ts (hunk 3 above)
Comment (LogvinovLeon) [inline ts/globals.d.ts]: Same - define the type

**Comment by fabioberger** [comment]
Re: caching already loaded JSON to memory, I don't think this is a worthwhile enhancement because I don't expect users to be frequently changing the versions of the docs they are looking at. Most will be using and building on one version, find it and browse within it.

File: ts/globals.d.ts [hunk 7]
```diff
@@ -36,3 +34,9 @@ declare function compareVersions(firstVersion: string, secondVersion: string): n
 declare module 'compare-versions' {
     export = compareVersions;
 }
+
+// semver-sort declarations
+declare module 'semver-sort' {
+    const desc: (versions: string[]) => string[];
+    const asc: (versions: string[]) => string[];
```
Comment (LogvinovLeon) [inline ts/globals.d.ts]: We don't use this one

File: ts/globals.d.ts [hunk 8]
```diff
@@ -40,3 +39,12 @@ declare module 'semver-sort' {
     const desc: (versions: string[]) => string[];
     const asc: (versions: string[]) => string[];
 }
+
+// xml-js declarations
+declare interface XML2JSONOpts {
```
Comment (LogvinovLeon) [inline ts/globals.d.ts]: Why do we need this comment?

**Pull request merged**